pkg/domain/stateDiffComponent.go

```
@@ -0,0 +1,187 @@
+			Version:               blueprintComponent.Version,
+			InstallationState:     blueprintComponent.TargetState,
+		}
+	}
+
+	return ComponentDiff{
+		Name:         componentName,
+		Expected:     expectedState,
+		Actual:       actualState,
+		NeededAction: getNextComponentAction(logger, expectedState, actualState),
+	}
+}
+
+func findComponentByName(components []Component, name string) (Component, bool) {
+	for _, component := range components {
+		if component.Name == name {
+			return component, true
+		}
+	}
+	return Component{}, false
+}
+
+func getNextComponentAction(logger logr.Logger, expected ComponentDiffState, actual ComponentDiffState) Action {
+	if expected.InstallationState == actual.InstallationState {
+		return decideOnEqualState(logger, expected, actual)
+	}
+
+	return decideOnDifferentState(logger, expected)
+}
+
+func decideOnEqualState(logger logr.Logger, expected ComponentDiffState, actual ComponentDiffState) Action {
+	switch expected.InstallationState {
+	case TargetStatePresent:
+		if expected.Version.IsNewerThan(actual.Version) {
+			return ActionUpgrade
+		}
+		if expected.Version.IsEqualTo(actual.Version) {
+			return ActionNone
+		}
+		return ActionDowngrade
+	case TargetStateAbsent:
+		return ActionNone
+	default:
+		logger.Info("Warning: Component has unexpected target state, deciding for no action",
+			"component", expected.InstallationState, "expected states", "")
+		return ActionNone
+
+	}
+}
+
+func decideOnDifferentState(logger logr.Logger, expected ComponentDiffState) Action {
+	// at this place, the actual state is always the opposite to the expected state so just follow the expected state.
+	switch expected.InstallationState {
+	case TargetStatePresent:
+		return ActionInstall
+	case TargetStateAbsent:
+		return ActionUninstall
+	default:
+		logger.Info("Warning: Component has unexpected installation state, deciding for no action",
+			"component", expected.InstallationState, "expected states", "")
+		return ActionNone
```

You pass this logger through five different methods before finally using it. 
Instead, it would probably be better to throw an error here and log it in the `stateDiffUseCase`.

---

- [x] done